src/python/pants/backend/python/tasks2/pytest_run.py

```diff
@@ -550,16 +557,136 @@ def _run_tests(self, targets):
       # A low-level test execution failure occurred before tests were run.
       raise TaskError()
 
-  def _do_run_tests(self, targets):
-    with self.invalidated(targets,
+  @staticmethod
+  def _vts_for_partition(invalidation_check):
+    return VersionedTargetSet.from_versioned_targets(invalidation_check.all_vts)
+
+  def check_artifact_cache_for(self, invalidation_check):
+    # We generate artifacts, namely junit.xml and coverage reports, that cover the full target set
+    # whether that is all targets in the context (`--fast`) or each target
+    # individually (`--no-fast`).
+    return [self._vts_for_partition(invalidation_check)]
+
+  @staticmethod
+  def _copy_dir(src, dst):
+    # NB: shutil.copytree requires dst not already exist, we tolerate that case.
+    safe_mkdir(dst)
+    for src_path, dirnames, filenames in safe_walk(src, topdown=True):
+      dst_path = os.path.join(dst, os.path.relpath(src_path, src))
+      for dirname in dirnames:
+        safe_mkdir(os.path.join(dst_path, dirname))
+      for filename in filenames:
+        dst_filename = os.path.join(dst_path, filename)
+        if os.path.exists(dst_filename):
+          os.unlink(dst_filename)
+        shutil.copy2(os.path.join(src_path, filename), dst_filename)
+
+  class Workdirs(datatype('Workdirs', ['root_dir'])):
```

Comment:
Making this an inner class seems unnecessary... private outer class (`class _Workdirs`) would make more sense to me, as it doesn't signal that these is any sort of coupling between the parent and child classes.